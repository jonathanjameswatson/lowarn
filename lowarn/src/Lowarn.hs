module Lowarn
  ( isUpdateAvailable,
  )
where

import Control.Concurrent (tryTakeMVar)
import Data.Maybe (isJust)
import Lowarn.Types (RuntimeData (..), UpdateSignal (..))

isUpdateAvailable :: RuntimeData a -> IO Bool
isUpdateAvailable (RuntimeData (UpdateSignal updateSignal) _) =
  isJust <$> tryTakeMVar updateSignal
