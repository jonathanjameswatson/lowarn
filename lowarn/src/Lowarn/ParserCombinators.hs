module Lowarn.ParserCombinators
  ( parsePackageName,
    parseProgramModuleName,
    readWithParser,
  )
where

import Control.Applicative (liftA2)
import Data.Char (isAsciiLower, isDigit)
import Data.List (intercalate)
import Data.Maybe (listToMaybe)
import Text.ParserCombinators.ReadP

parsePackageWord :: ReadP String
parsePackageWord =
  concat
    <$> sequence
      [ munch isDigit,
        munch1 isAsciiLower,
        munch (liftA2 (||) isAsciiLower isDigit)
      ]

parsePackageWordSequence :: Char -> ReadP String
parsePackageWordSequence separator =
  intercalate [separator] <$> sepBy1 parsePackageWord (char separator)

parsePackageName :: ReadP String
parsePackageName = parsePackageWordSequence '-'

parseProgramModuleName :: ReadP String
parseProgramModuleName = parsePackageWordSequence '_'

readWithParser :: ReadP a -> String -> Maybe a
readWithParser parser = fmap fst . listToMaybe . readP_to_S (parser <* eof)
