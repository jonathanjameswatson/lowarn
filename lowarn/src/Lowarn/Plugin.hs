-- |
-- Module                  : Lowarn.Runtime
-- SPDX-License-Identifier : MIT
-- Stability               : experimental
-- Portability             : non-portable (POSIX, GHC)
module Lowarn.Plugin (plugin) where

import GHC.Plugins

plugin :: Plugin
plugin =
  defaultPlugin
    { installCoreToDos = install
    }

install :: [CommandLineOption] -> [CoreToDo] -> CoreM [CoreToDo]
install _ todo = do
  putMsgS "Lowarn plugin enabled."
  return todo
