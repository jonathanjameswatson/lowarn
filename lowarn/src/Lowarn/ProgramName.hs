module Lowarn.ProgramName
  ( ProgramName (..),
    ofVersionModuleName,
    ofTransformerModuleName,
    toVersionModuleName,
    toTransformerModuleName,
  )
where

import Text.Regex.TDFA

newtype ProgramName = ProgramName
  { unProgramName :: String
  }
  deriving (Eq, Show, Read)

replace :: Eq a => a -> a -> [a] -> [a]
replace a b = map $ \c -> if c == a then b else c

replaceUnderscoresWithHyphens :: String -> String
replaceUnderscoresWithHyphens = replace '_' '-'

replaceHyphensWithUnderscores :: String -> String
replaceHyphensWithUnderscores = replace '-' '_'

ofModuleName :: String -> String -> Maybe String
ofModuleName prefix moduleName = case submatches of
  [programName] -> Just $ replaceUnderscoresWithHyphens programName
  _ -> Nothing
  where
    (_, _, _, submatches) =
      moduleName =~ ("\\`" <> prefix <> "_(.+)\\'") ::
        (String, String, String, [String])

ofVersionModuleName :: String -> Maybe String
ofVersionModuleName = ofModuleName "Version"

ofTransformerModuleName :: String -> Maybe String
ofTransformerModuleName = ofModuleName "Transformer"

toModuleName :: String -> ProgramName -> String
toModuleName prefix =
  ((prefix <> "_") <>)
    . replaceHyphensWithUnderscores
    . unProgramName

toVersionModuleName :: ProgramName -> String
toVersionModuleName = toModuleName "Version"

toTransformerModuleName :: ProgramName -> String
toTransformerModuleName = toModuleName "Transformer"
