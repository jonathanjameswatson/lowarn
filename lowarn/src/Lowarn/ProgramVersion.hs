{-# LANGUAGE LambdaCase #-}

module Lowarn.ProgramVersion
  ( ProgramVersion (..),
    showWithDots,
    showWithLetters,
    parseWithDots,
    parseWithLetters,
  )
where

import Data.Char (isDigit)
import Data.List (intercalate)
import Data.List.NonEmpty (NonEmpty ((:|)), toList)
import Text.ParserCombinators.ReadP

newtype ProgramVersion = ProgramVersion
  { unProgramVersion :: NonEmpty Int
  }
  deriving (Eq, Ord, Show, Read)

showWithSeparator :: String -> ProgramVersion -> String
showWithSeparator separator = intercalate separator . map show . toList . unProgramVersion

showWithDots :: ProgramVersion -> String
showWithDots = showWithSeparator "."

showWithLetters :: ProgramVersion -> String
showWithLetters = ("v" <>) . showWithSeparator "v"

parseWithDots :: ReadP ProgramVersion
parseWithDots =
  sepBy1 (read <$> munch1 isDigit) (char '.') >>= \case
    [] -> pfail
    (x : xs) -> return $ ProgramVersion (x :| xs)

parseWithLetters :: ReadP ProgramVersion
parseWithLetters =
  char 'v' *> sepBy1 (read <$> munch1 isDigit) (char 'v') >>= \case
    [] -> pfail
    (x : xs) -> return $ ProgramVersion (x :| xs)
