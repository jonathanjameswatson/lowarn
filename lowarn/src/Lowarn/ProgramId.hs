module Lowarn.ProgramId
  ( ProgramId (..),
    showProgramId,
    parseProgramId,
    ofEntryPointPackageName,
    ofTransformerPackageName,
    toEntryPointPackageName,
    toTransformerPackageName,
  )
where

import Control.Monad (void)
import Data.Maybe (listToMaybe)
import Lowarn.ProgramName (ProgramName (..))
import Lowarn.ProgramVersion
  ( ProgramVersion,
    parseWithDots,
    parseWithLetters,
    showWithDots,
    showWithLetters,
  )
import Text.ParserCombinators.ReadP
import Text.Regex.TDFA

data ProgramId = ProgramId
  { _programName :: ProgramName,
    _programVersion :: ProgramVersion
  }
  deriving (Eq, Show, Read)

showProgramId :: ProgramId -> String
showProgramId (ProgramId programName programVersion) =
  unProgramName programName <> "-" <> showWithDots programVersion

parseProgramId :: ReadP ProgramId
parseProgramId = do
  name <- many1 (satisfy $ const True)
  void $ char '-'
  ProgramId (ProgramName name) <$> parseWithDots

ofPackageName :: String -> String -> Maybe ProgramId
ofPackageName prefix packageName = case submatches of
  programName : programVersion : _ ->
    ProgramId
      (ProgramName programName)
      <$> (fst <$> listToMaybe (readP_to_S parseWithLetters programVersion))
  _ -> Nothing
  where
    (_, _, _, submatches) =
      packageName =~ ("\\`lowarn-" <> prefix <> "-(.+)-((v[[:digit:]]+)+)\\'") ::
        (String, String, String, [String])

ofEntryPointPackageName :: String -> Maybe ProgramId
ofEntryPointPackageName = ofPackageName "entry-point"

ofTransformerPackageName :: String -> Maybe ProgramId
ofTransformerPackageName = ofPackageName "transformer"

toPackageName :: String -> ProgramId -> String
toPackageName prefix (ProgramId programName programVersion) =
  "lowarn-"
    <> prefix
    <> "-"
    <> unProgramName programName
    <> "-"
    <> showWithLetters programVersion

toEntryPointPackageName :: ProgramId -> String
toEntryPointPackageName = toPackageName "entry-point"

toTransformerPackageName :: ProgramId -> String
toTransformerPackageName = toPackageName "transformer"
