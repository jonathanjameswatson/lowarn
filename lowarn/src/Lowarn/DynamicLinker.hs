{-# LANGUAGE GeneralizedNewtypeDeriving #-}

-- |
-- Module                  : Lowarn.Runtime
-- SPDX-License-Identifier : MIT
-- Stability               : experimental
-- Portability             : non-portable (POSIX, GHC)
--
-- Module for interacting with GHC to link modules and load their entities.
module Lowarn.DynamicLinker
  ( Linker,
    runLinker,
    load,
    liftIO,
  )
where

import Control.Monad (void)
import Control.Monad.IO.Class (MonadIO)
import GHC hiding (load, moduleName, unitState)
import GHC.Driver.Monad
import GHC.Driver.Session hiding (unitState)
import qualified GHC.Driver.Session as Session
import GHC.Driver.Ways
import GHC.Paths (libdir)
import GHC.Runtime.Interpreter
import GHC.Runtime.Linker
import GHC.Types.Name
import GHC.Types.Unique
import GHC.Unit hiding (moduleName)
import Unsafe.Coerce (unsafeCoerce)

-- | Monad for linking modules from the package database and accessing their
-- exported entities.
newtype Linker a = Linker
  { unLinker :: Ghc a
  }
  deriving (Functor, Applicative, Monad, MonadIO)

-- | Run a linker.
runLinker :: Linker a -> IO a
runLinker linker =
  defaultErrorHandler defaultFatalMessager defaultFlushOut $
    runGhc (Just libdir) $ do
      flags <- getSessionDynFlags
      void $
        setSessionDynFlags $
          addWay' WayDyn $
            flags {ghcMode = CompManager, ghcLink = LinkDynLib}
      liftIO . initDynLinker =<< getSession
      unLinker linker

-- | Action that gives an entity exported by a module in the package database,
-- which is linked if it hasn't already been.
load ::
  -- | The name of the module to find. @Nothing@ is given by the action if the
  -- module cannot be found.
  String ->
  -- | The name of the entity to take from the module.
  String ->
  Linker (Maybe a)
load moduleName' symbol = Linker $ do
  flags <- getSessionDynFlags
  session <- getSession

  let moduleName = mkModuleName moduleName'

  liftIO $
    lookupUnitInfo flags moduleName
      >>= maybe
        (return Nothing)
        ( \unitInfo -> liftIO $ do
            let unit = mkUnit unitInfo
            let module_ = mkModule unit moduleName
            let name =
                  mkExternalName
                    (mkBuiltinUnique 0)
                    module_
                    (mkVarOcc symbol)
                    noSrcSpan

            linkModule session module_

            value <- withInterp session $
              \interp -> getHValue session name >>= wormhole interp
            return $ Just $ unsafeCoerce value
        )

lookupUnitInfo :: DynFlags -> ModuleName -> IO (Maybe UnitInfo)
lookupUnitInfo flags moduleName = do
  case exposedModulesAndPackages of
    [] -> do
      putStrLn $ "Can't find module " <> moduleNameString moduleName
      return Nothing
    (_, unitInfo) : _ -> return $ Just unitInfo
  where
    unitState = Session.unitState flags
    modulesAndPackages = lookupModuleInAllUnits unitState moduleName
    exposedModulesAndPackages = filter (unitIsExposed . snd) modulesAndPackages
