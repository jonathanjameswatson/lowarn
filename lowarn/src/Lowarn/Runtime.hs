{-# LANGUAGE GeneralisedNewtypeDeriving #-}

-- |
-- Module                  : Lowarn.Runtime
-- SPDX-License-Identifier : MIT
-- Stability               : experimental
-- Portability             : non-portable (POSIX, GHC)
--
-- Module for creating and interacting with DSU runtimes.
module Lowarn.Runtime
  ( -- * Runtime creation
    Runtime,
    runRuntime,
    loadProgram,
    liftLinker,
    liftIO,

    -- * Runtime interaction
    Program (..),
    RuntimeData,
    isUpdateAvailable,
    lastState,
  )
where

import Control.Concurrent (MVar, newEmptyMVar, tryPutMVar, tryTakeMVar)
import Control.Monad (void)
import Control.Monad.IO.Class (MonadIO)
import Control.Monad.Trans.Reader (ReaderT, ask, runReaderT)
import Data.Maybe (isJust)
import Lowarn.DynamicLinker (Linker, liftIO, load, runLinker)
import System.Posix.Signals (Handler (Catch), installHandler, sigUSR2)

newtype UpdateSignal = UpdateSignal
  { unUpdateSignal :: MVar ()
  }

-- | Monad for DSU runtimes that load versions of programs.
newtype Runtime a = Runtime
  { unRuntime :: ReaderT UpdateSignal IO a
  }
  deriving (Functor, Applicative, Monad, MonadIO)

-- | Type for exposing functions that are used by the runtime to load a program.
data Program a b = Program
  { -- | Function to start the program, given data injected by the runtime.
    _entryPoint :: RuntimeData a -> IO a,
    -- | Function to convert old state to new state, if this can be done.
    _transformer :: b -> IO (Maybe a)
  }

-- | Type for accessing data injected by the runtime.
data RuntimeData a = RuntimeData
  { _updateSignal :: UpdateSignal,
    _updateInfo :: Maybe (UpdateInfo a)
  }

data UpdateInfo a = UpdateInfo
  { _label :: Maybe String,
    _lastState :: a
  }

-- | Run a runtime.
runRuntime :: Runtime a -> IO a
runRuntime runtime = do
  updateSignal <- newEmptyMVar
  previousSignalHandler <-
    installHandler
      sigUSR2
      (Catch (void $ tryPutMVar updateSignal ()))
      Nothing
  output <- runReaderT (unRuntime runtime) $ UpdateSignal updateSignal
  liftIO $ void $ installHandler sigUSR2 previousSignalHandler Nothing
  return output

-- | Action that loads and runs a given program, producing the final state of
-- the program when it finishes.
--
-- The program is given as the name of a module that exports a value
-- @program :: 'Program' b a@. This module should be in the package database.
--
-- The program is also given data representing state from the previous version
-- of the program.
loadProgram ::
  -- | The name of the module that includes the program.
  String ->
  -- | State from the previous version of the program.
  a ->
  Runtime b
loadProgram moduleName lastState_ = Runtime $ do
  updateSignal <- ask
  liftIO $
    runLinker (load moduleName "program")
      >>= maybe
        (error ("Loading " <> moduleName <> " failed"))
        ( \program -> do
            updateInfo <-
              fmap (UpdateInfo Nothing)
                <$> _transformer program lastState_
            _entryPoint program $ RuntimeData updateSignal updateInfo
        )

-- | Lift a computation from the 'Linker' monad.
liftLinker :: Linker a -> Runtime a
liftLinker = Runtime . liftIO . runLinker

-- | Return @True@ if the runtime has a program update that can be applied.
isUpdateAvailable :: RuntimeData a -> IO Bool
isUpdateAvailable =
  (fmap isJust . tryTakeMVar) . (unUpdateSignal . _updateSignal)

-- | Return the transformed state of the last version of the program, if there
-- was a previous version of the program and the state was able to be
-- transformed.
lastState :: RuntimeData a -> Maybe a
lastState = fmap _lastState . _updateInfo
