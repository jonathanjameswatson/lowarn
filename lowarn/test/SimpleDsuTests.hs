{-# LANGUAGE TemplateHaskellQuotes #-}

module SimpleDsuTests (goldenTests) where

import Control.Monad (void)
import DsuTest
  ( dsuGoldenTest,
    inputLine,
    outputLine,
    outputLines,
    updateProgram,
  )
import Following
import Lowarn.Runtime (Runtime, loadTransformer, loadVersion)
import System.IO (Handle)
import Test.Tasty (TestTree, testGroup)

getExampleRuntime :: (Handle, Handle) -> Runtime ()
getExampleRuntime handles =
  void $
    loadTransformer followingTransformerId_0_1 handles
      >>= loadVersion followingVersionId_1
      >>= loadTransformer followingTransformerId_1_2
      >>= loadVersion followingVersionId_2
      >>= loadTransformer followingTransformerId_2_3
      >>= loadVersion followingVersionId_3

successfulChain :: TestTree
successfulChain =
  dsuGoldenTest
    (show 'successfulChain)
    getExampleRuntime
    dsuTest
    10000000
  where
    dsuTest = do
      _ <- outputLines 3
      inputLine "A"
      _ <- outputLines 4
      updateProgram
      inputLine "B"

      _ <- outputLines 5
      inputLine "C"
      _ <- outputLines 1
      inputLine "1234"
      _ <- outputLines 6
      updateProgram
      inputLine "D"
      _ <- outputLine
      inputLine "2345"

      _ <- outputLines 7
      inputLine "E#3456"
      _ <- outputLines 8
      updateProgram
      inputLine "F#4567"

duplicatedUpdateSignal :: TestTree
duplicatedUpdateSignal =
  dsuGoldenTest
    (show 'duplicatedUpdateSignal)
    getExampleRuntime
    dsuTest
    10000000
  where
    dsuTest = do
      _ <- outputLines 3
      inputLine "A"
      _ <- outputLines 4
      updateProgram
      updateProgram
      updateProgram
      inputLine "B"

      _ <- outputLines 5
      inputLine "C"
      _ <- outputLines 1
      inputLine "1234"
      _ <- outputLines 6
      return ()

goldenTests :: TestTree
goldenTests =
  testGroup
    "Simple DSU runtime golden tests"
    [ successfulChain,
      duplicatedUpdateSignal
    ]
