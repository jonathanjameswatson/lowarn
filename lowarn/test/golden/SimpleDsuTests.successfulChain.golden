Output: Blocked users:
Output: ------
Output: Input username of user to block:
 Input: A
Output: Blocked users:
Output: A
Output: ------
Output: Input username of user to block:
  Info: Update signal sent.
 Input: B
Output: Blocked users:
Output: A#8153
Output: B#9481
Output: ------
Output: Input nickname of user to block:
 Input: C
Output: Input user ID of user to block:
 Input: 1234
Output: Blocked users:
Output: A#8153
Output: B#9481
Output: C#1234
Output: ------
Output: Input nickname of user to block:
  Info: Update signal sent.
 Input: D
Output: Input user ID of user to block:
 Input: 2345
Output: Blocked users:
Output: A#8153
Output: B#9481
Output: C#1234
Output: D#2345
Output: ------
Output: Input tag of user to block:
 Input: E#3456
Output: Blocked users:
Output: A#8153
Output: B#9481
Output: C#1234
Output: D#2345
Output: E#3456
Output: ------
Output: Input tag of user to block:
  Info: Update signal sent.
 Input: F#4567
